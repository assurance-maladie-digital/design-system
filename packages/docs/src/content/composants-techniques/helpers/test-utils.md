---
title: testUtils
description: Le helper `testUtils` contient un ensemble de fonctions utilitaires pour les tests unitaires.
---

<doc-tabs>

<doc-tab-item label="Utilisation">

<doc-alert type="info">

Le helper `testUtils` exporte également toutes les fonctions du package [Vue Test Utils](https://vue-test-utils.vuejs.org/).

</doc-alert>

### applyGlobalConfig

La fonction `applyGlobalConfig` permet d’appliquer la configuration globale pour les tests, elle est utilisée dans le fichier `setup.ts`.

```ts
import { applyGlobalConfig } from '@cnamts/vue-dot/src/helpers/testUtils';

applyGlobalConfig();
```

### installGlobalPlugins

La fonction `installGlobalPlugins` permet d’installer les plugins [Vue Input Facade](https://ronaldjerez.github.io/vue-input-facade/), [Vue Meta](https://vue-meta.nuxtjs.org/), [Vuex](https://vuex.vuejs.org/), [Vuetify](https://v2.vuetifyjs.com/) et Vue Dot pour les tests.

```ts
import { createLocalVue, installGlobalPlugins } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installGlobalPlugins(localVue);
```

### installRouter

La fonction `installRouter` permet d’installer [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createLocalVue, installRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installRouter(localVue);
```

### createRouter

La fonction `createRouter` permet de créer une instance de [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const routes = [];

const router = createRouter({
	routes
});
```

### createStore

La fonction `createStore` permet de créer une instance de [Vuex](https://vuex.vuejs.org/) pour les tests.

```ts
import { createStore } from '@cnamts/vue-dot/src/helpers/testUtils';

const store = createStore({
	state: {}
});
```

### createVuetifyInstance

La fonction `createVuetifyInstance` permet de créer une instance de [Vuetify](https://v2.vuetifyjs.com/) pour les tests.

```ts
import { createVuetifyInstance } from '@cnamts/vue-dot/src/helpers/testUtils';

const vuetify = createVuetifyInstance();
```

### mockVFormRef

La fonction `mockVFormRef` permet de mocker la référence d’un composant [VForm](https://v2.vuetifyjs.com/en/components/forms/) pour les tests.

```ts
import { mockVFormRef } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	$refs: {
		form: mockVFormRef(true)
	}
};
```

### mockTranslations

La fonction `mockTranslations` permet de mocker les traductions pour les tests.

```ts
import { mockTranslations } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	...mockTranslations<string | string[]>({
		'path.to.title': 'Titre',
		'path.to.items': [
			'Élément 1',
			'Élément 2',
			'Élément 3'
		]
	})
};
```

</doc-tab-item>

<doc-tab-item label="API">
<doc-api name="helpers/test-utils"></doc-api>
</doc-tab-item>

</doc-tabs>
