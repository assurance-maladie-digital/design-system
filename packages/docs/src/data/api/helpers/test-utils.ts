import { Api } from '~/types';

export const api: Api = {
	applyGlobalConfig: {
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installGlobalPlugins: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installRouter: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	createRouter: {
		arguments: [
			{
				name: 'options',
				description: 'Router options',
				type: 'RouterOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'VueRouter',
			},
		],
	},
	createStore: {
		arguments: [
			{
				name: 'options',
				description: 'Store options',
				type: 'StoreOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'Store',
			},
		],
	},
	createVuetifyInstance: {
		arguments: [
			{
				name: 'preset',
				description: 'Vuetify preset',
				type: 'Partial<UserVuetifyPreset>',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'Vuetify',
			},
		],
	},
	mockVFormRef: {
		arguments: [
			{
				name: 'isValid',
				description: 'Is valid',
				type: 'boolean',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	mockTranslations: {
		arguments: [
			{
				name: 'translations',
				description: 'Translations',
				type: 'IndexedObject',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
};
