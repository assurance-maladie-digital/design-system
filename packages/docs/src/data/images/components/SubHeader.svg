<svg width="541" height="357" viewBox="0 0 541 357" fill="none" xmlns="http://www.w3.org/2000/svg">
<rect width="541" height="357" fill="#E7ECF5"/>
<rect x="63" y="88" width="415" height="181" fill="white"/>
<rect opacity="0.5" x="116" y="114" width="69" height="20" rx="10" fill="#CED9EB"/>
<path d="M94.3333 129.667L95.2733 128.727L91.5533 125L105.667 125L105.667 123.667L91.5533 123.667L95.28 119.947L94.3333 119L89 124.333L94.3333 129.667Z" fill="#0C419A"/>
<rect opacity="0.5" x="89" y="151" width="112" height="20" rx="10" fill="#CED9EB"/>
<rect opacity="0.5" x="224" y="151" width="230" height="20" rx="10" fill="#CED9EB"/>
<rect opacity="0.5" x="224" y="187" width="230" height="20" rx="10" fill="#CED9EB"/>
<rect opacity="0.5" x="224" y="224" width="230" height="20" rx="10" fill="#CED9EB"/>
</svg>
