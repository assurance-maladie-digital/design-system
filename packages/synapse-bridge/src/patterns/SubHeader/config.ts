export const config = {
	sheet: {
		color: 'secondary'
	},
	backBtn: {
		size: 'small',
		variant: 'text',
		class: 'font-weight-regular white--text px-1'
	}
};
