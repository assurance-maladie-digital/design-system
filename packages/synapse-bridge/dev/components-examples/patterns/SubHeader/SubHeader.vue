<template>
	<SubHeader
		back-btn-text="Back"
		sub-title-text="1 69 08 75 125 456 75"
		title-text="Paul Dupont"
	/>

	<br />

	<SubHeader
		:data-list-group-items="dataListGroupItems"
		title-text="Paul Dupont"
		sub-title-text="1 69 08 75 125 456 75"
		render-html-value
	/>

	<br />

	<SubHeader
		title-text="Paul Dupont"
		sub-title-text="1 69 08 75 125 456 75"
	>
		<template #additional-informations>
			<VSpacer />

			<p class="white--text mt-8 mb-0">
				Profil complété à 50%
			</p>

			<VProgressLinear
				:model-value="50"
				color="#fff"
				height="8px"
				class="mt-2 mb-1"
				background-color="#fff"
				background-opacity=".24"
			/>
		</template>
	</SubHeader>

	<br />

	<SubHeader
		:data-list-group-items="dataListGroupItems"
		:loading="loading"
		title-text="Paul Dupont"
		sub-title-text="1 69 08 75 125 456 75"
	>
		<template #additional-informations>
			<VSpacer />

			<VFadeTransition mode="out-in">
				<div
					v-if="loading"
					key="progress-skeleton"
				>
					<HeaderLoading
						class="mt-8"
						height="24"
						width="160"
						dark
					/>

					<HeaderLoading
						class="mt-2 mb-1"
						height="8"
						width="100%"
						tile
						dark
					/>
				</div>

				<div
					v-else
					key="progress-data"
				>
					<p class="white--text mt-8 mb-0">
						Profil complété à 50%
					</p>

					<VProgressLinear
						:model-value="50"
						color="#fff"
						height="8px"
						class="mt-2 mb-1"
						background-color="#fff"
						background-opacity=".24"
					/>
				</div>
			</VFadeTransition>
		</template>
	</SubHeader>

	<VSwitch
		v-model="loading"
		label="Chargement"
		hide-details
	/>
</template>

<script lang="ts">
import { defineComponent } from "vue";

import SubHeader from "@/patterns/SubHeader/SubHeader.vue";
import HeaderLoading from "@/elements/HeaderLoading/HeaderLoading.vue";
import { DataListGroupItems } from '@/patterns/DataListGroup/types';

export default defineComponent({
	components: {
		SubHeader,
		HeaderLoading
	},
	data() {
		return {
			loading: true,
			dataListGroupItems: [
				{
					title: 'Informations patient',
					itemsNumberLoading: 2,
					headingLoading: true,
					items: [
						{
							key: 'Date de naissance',
							value: '24/09/1970'
						},
						{
							key: 'Adresse',
							value: '50 Avenue du Professeur André Lemierre<br>75020 Paris'
						}
					]
				},
				{
					title: 'Médecin traitant',
					itemsNumberLoading: 2,
					headingLoading: true,
					items: [
						{
							key: 'Nom du praticien',
							value: 'Gérard Leblanc'
						},
						{
							key: 'N° RPPS',
							value: 'XXXXX'
						}
					]
				},
				{
					title: 'Autres informations',
					itemsNumberLoading: 1,
					headingLoading: true,
					items: [
						{
							key: 'Dernière modification',
							value: '04/06/2020'
						}
					]
				}
			] satisfies DataListGroupItems
		};
	}
});
</script>
