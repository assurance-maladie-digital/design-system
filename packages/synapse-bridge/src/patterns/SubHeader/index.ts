import SubHeader from './SubHeader.vue'

export default SubHeader
