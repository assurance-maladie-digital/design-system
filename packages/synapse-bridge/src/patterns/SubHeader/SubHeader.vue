<script lang="ts">
import { defineComponent } from "vue";
import type { PropType } from "vue";

import { config } from "./config";
import { locales } from "./locales";

import { customizable } from "@/mixins/customizable";
import { Widthable } from "@/mixins/widthable";

import DataListGroup from "@/patterns/DataListGroup";
import {
	DataListActionEvent,
	DataListGroupItems,
} from "@/patterns/DataListGroup/types";

import { mdiKeyboardBackspace } from "@mdi/js";
import HeaderLoading from '@/elements/HeaderLoading';
import { VSkeletonLoader } from 'vuetify/components/VSkeletonLoader';

export default defineComponent({
	components: {
		DataListGroup,
		HeaderLoading,
		VSkeletonLoader
	},
	mixins: [customizable(config), Widthable],
	props: {
		hideBackBtn: {
			type: Boolean,
			default: false,
		},
		backBtnText: {
			type: String,
			default: locales.backBtnText,
		},
		titleText: {
			type: String,
			default: undefined,
		},
		subTitleText: {
			type: String,
			default: undefined,
		},
		dataListGroupItems: {
			type: Array as PropType<DataListGroupItems | undefined>,
			default: undefined,
		},
		loading: {
			type: Boolean,
			default: false,
		},
		renderHtmlValue: {
			type: Boolean,
			default: false,
		},
	},
	emits: ["click:list-item", "back"],
	data() {
		return {
			backArrowIcon: mdiKeyboardBackspace,
			locales,
		};
	},
	computed: {
		fadeWhite(): string {
			return 'rgba(255, 255, 255, .7)';
		}
	},
	methods: {
		emitItemAction(eventValue: DataListActionEvent): void {
			this.$emit('click:list-item', eventValue);
		}
	}
});
</script>

<template>
	<VSheet
		v-bind="options.sheet"
		:style="widthStyles"
		class="vd-sub-header white--text py-6 px-8"
	>
		<slot name="back-btn">
			<VFadeTransition v-if="!hideBackBtn" mode="out-in">
				<VSkeletonLoader
					v-if="loading"
					max-height="28"
					type="button"
					color="secondary"
					class="vd-subheader-loading mb-4"
				/>
				<VBtn
					v-else
					v-bind="options.backBtn"
					class="vd-sub-header-back-btn mb-1"
					@click="$emit('back')"
				>
					<slot name="back-btn-icon">
						<VIcon class="mr-2">
							{{ backArrowIcon }}
						</VIcon>
					</slot>

					{{ backBtnText }}
				</VBtn>
			</VFadeTransition>
		</slot>

		<div class="vd-sub-header-content d-flex justify-space-between">
			<div
				class="vd-sub-header-informations d-flex flex-column flex-shrink-0 mr-10"
			>
				<slot name="title">
					<VFadeTransition mode="out-in">
						<HeaderLoading
							v-if="loading"
							width="300"
							height="2rem"
							color="secondary"
						/>

						<h2
							v-else-if="titleText"
							class="text-h5 font-weight-bold"
						>
							{{ titleText }}
						</h2>
					</VFadeTransition>
				</slot>

				<slot name="sub-title">
					<VFadeTransition v-if="subTitleText" mode="out-in">
						<HeaderLoading
							v-if="loading"
							class="mt-1"
							width="250"
							height="2rem"
							color="secondary"
						/>

						<p
							v-else
							class="text-h6 font-weight-bold mt-1 mb-0"
							:style="{ color: fadeWhite }"
						>
							{{ subTitleText }}
						</p>
					</VFadeTransition>
				</slot>

				<slot name="additional-informations" />
			</div>

			<slot name="right-content">
				<DataListGroup
					v-if="dataListGroupItems"
					:items="dataListGroupItems"
					:loading="loading"
					:render-html-value="renderHtmlValue"
					item-width="auto"
					class="flex-nowrap flex-shrink-0"
					@click:list-item="emitItemAction"
				/>
			</slot>
		</div>
	</VSheet>
</template>

<style lang="scss" scoped>
@use '@cnamts/design-tokens/dist/tokens';

.vd-sub-header {
	overflow-x: auto;
}

.vd-sub-header-back-btn {
	margin: 0 -6px;
}

.vd-data-list-group,
.vd-sub-header-informations {
	// Don't take all available space
	max-width: none;
}

.vd-data-list-group :deep(.vd-data-list) {
	max-width: 200px;

	// Apply margin right to avoid empty
	// space on smaller screens
	&:not(:last-child) {
		margin-right: 80px !important;
	}

	.vd-key {
		display: inline-block;
		font-size: 0.75rem !important;
	}

	.vd-data-list-item-label {
		color: rgba(255, 255, 255, 0.7) !important;
	}

	.vd-data-list-item-action-btn {
		color: #fff !important;
	}
}

.vd-subheader-loading :deep() {
	background: transparent;
	.v-skeleton-loader__button {
		margin: 0;
		min-height: 28px;
		height: auto;
		background: rgba(white, var(--v-border-opacity));
	}
}
</style>
