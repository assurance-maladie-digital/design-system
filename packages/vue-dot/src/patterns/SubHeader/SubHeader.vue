<template>
	<VSheet
		v-bind="options.sheet"
		:style="widthStyles"
		class="vd-sub-header white--text py-6 px-8"
	>
		<slot name="back-btn">
			<VFadeTransition
				v-if="!hideBackBtn"
				mode="out-in"
			>
				<VSkeletonLoader
					v-if="loading"
					height="28"
					type="button"
					width="100"
					class="mb-1"
					dark
				/>

				<VBtn
					v-else
					v-bind="options.backBtn"
					class="vd-sub-header-back-btn mb-1"
					@click="$emit('back')"
				>
					<slot name="back-btn-icon">
						<VIcon class="mr-2">
							{{ backArrowIcon }}
						</VIcon>
					</slot>

					{{ backBtnText }}
				</VBtn>
			</VFadeTransition>
		</slot>

		<div class="vd-sub-header-content d-flex justify-space-between">
			<div class="vd-sub-header-informations d-flex flex-column flex-shrink-0 mr-10">
				<slot name="title">
					<VFadeTransition mode="out-in">
						<HeaderLoading
							v-if="loading"
							width="300"
							height="2rem"
							dark
						/>

						<h2
							v-else-if="titleText"
							class="text-h5 font-weight-bold"
						>
							{{ titleText }}
						</h2>
					</VFadeTransition>
				</slot>

				<slot name="sub-title">
					<VFadeTransition
						v-if="subTitleText"
						mode="out-in"
					>
						<HeaderLoading
							v-if="loading"
							class="mt-1"
							width="250"
							height="2rem"
							dark
						/>

						<p
							v-else
							class="text-h6 font-weight-bold mt-1 mb-0"
							:style="{ color: fadeWhite }"
						>
							{{ subTitleText }}
						</p>
					</VFadeTransition>
				</slot>

				<slot name="additional-informations" />
			</div>

			<slot name="right-content">
				<VThemeProvider dark>
					<DataListGroup
						v-if="dataListGroupItems"
						:items="dataListGroupItems"
						:loading="loading"
						:render-html-value="renderHtmlValue"
						item-width="auto"
						class="flex-nowrap flex-shrink-0"
						@click:list-item="emitItemAction"
					/>
				</VThemeProvider>
			</slot>
		</div>
	</VSheet>
</template>

<script lang="ts">
	import Vue, { PropType } from 'vue';
	import Component, { mixins } from 'vue-class-component';

	import { config } from './config';
	import { locales } from './locales';

	import { customizable } from '../../mixins/customizable';
	import { Widthable } from '../../mixins/widthable';

	import DataListGroup from '../../patterns/DataListGroup';
	import { DataListActionEvent, DataListGroupItems } from '../DataListGroup/types';

	import { mdiKeyboardBackspace } from '@mdi/js';

	const Props = Vue.extend({
		props: {
			hideBackBtn: {
				type: Boolean,
				default: false
			},
			backBtnText: {
				type: String,
				default: locales.backBtnText
			},
			titleText: {
				type: String,
				default: undefined
			},
			subTitleText: {
				type: String,
				default: undefined
			},
			dataListGroupItems: {
				type: Array as PropType<DataListGroupItems | undefined>,
				default: undefined
			},
			loading: {
				type: Boolean,
				default: false
			},
			renderHtmlValue: {
				type: Boolean,
				default: false
			}
		}
	});

	const MixinsDeclaration = mixins(Props, customizable(config), Widthable);

	@Component({
		components: {
			DataListGroup
		}
	})
	export default class SubHeader extends MixinsDeclaration {
		backArrowIcon = mdiKeyboardBackspace;

		get fadeWhite(): string {
			return 'rgba(255, 255, 255, .7)';
		}

		emitItemAction(eventValue: DataListActionEvent): void {
			this.$emit('click:list-item', eventValue);
		}
	}
</script>

<style lang="scss" scoped>
	.vd-sub-header {
		overflow-x: auto;
	}

	.vd-sub-header-back-btn {
		margin: 0 -6px;
	}

	.vd-data-list-group,
	.vd-sub-header-informations {
		// Don't take all available space
		max-width: none;
	}

	.vd-data-list-group :deep(.vd-data-list) {
		max-width: 200px;

		// Apply margin right to avoid empty
		// space on smaller screens
		&:not(:last-child) {
			margin-right: 80px !important;
		}

		.vd-key {
			display: inline-block;
			font-size: .75rem !important;
		}

		.vd-data-list-item-label {
			color: rgba(255, 255, 255, .7) !important;
		}

		.vd-data-list-item-action-btn {
			color: #fff !important;
		}
	}
</style>
