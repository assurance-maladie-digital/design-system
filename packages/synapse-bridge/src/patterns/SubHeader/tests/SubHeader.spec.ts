import { describe, it, expect } from "vitest";
import { shallowMount } from "@vue/test-utils";

import SubHeader from "../";

import { dataListGroupItems } from "./data/subHeader";

describe("SubHeader", () => {
	it("renders correctly", () => {
		const wrapper = shallowMount(SubHeader, {
			propsData: {
				titleText: "Test",
			},
		});

		expect(wrapper).toMatchSnapshot();
	});

	it("renders loading state correctly", async () => {
		const wrapper = shallowMount(
			SubHeader,
			{
				propsData: {
					titleText: "Test",
					loading: true,
					dataListGroupItems,
				},
			}
		);

		expect(wrapper).toMatchSnapshot();
	});

	it("returns true when the header is not fixed", () => {
		const wrapper = shallowMount(SubHeader, {
			propsData: {
				titleText: "Test",
			},
		});

		expect(wrapper.vm.fadeWhite).toBe('rgba(255, 255, 255, .7)');
	});

	it("emits itemAction event when called", () => {
		const wrapper = shallowMount(SubHeader, {
			propsData: {
				titleText: "Test",
			},
		});

		wrapper.vm.emitItemAction({
			dataListIndex: 0,
			itemIndex: 0
		});

		expect(wrapper.emitted('click:list-item')).toBeTruthy();
	});
});
