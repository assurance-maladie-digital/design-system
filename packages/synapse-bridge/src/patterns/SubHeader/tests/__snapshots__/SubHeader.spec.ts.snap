// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SubHeader > renders correctly 1`] = `<v-sheet-stub data-v-04835fcd="" color="secondary" class="vd-sub-header white--text py-6 px-8" border="false" style="width: 100%;" tile="false" tag="div"></v-sheet-stub>`;

exports[`SubHeader > renders loading state correctly 1`] = `<v-sheet-stub data-v-04835fcd="" color="secondary" class="vd-sub-header white--text py-6 px-8" border="false" style="width: 100%;" tile="false" tag="div"></v-sheet-stub>`;
